ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/TextEmbeddingModel.java

```diff
@@ -36,6 +46,52 @@ public ModelTensorOutput predict(String modelId, MLInput mlInput) throws Transla
         return new ModelTensorOutput(tensorOutputs);
     }
 
+    private boolean isAsymmetricModel(MLAlgoParams mlParams) {
+        if (mlParams instanceof AsymmetricTextEmbeddingParameters) {
+            // Check for the necessary prefixes in modelConfig
+            if (modelConfig == null
+                || ((TextEmbeddingModelConfig) modelConfig).getPassagePrefix() == null
+                    && ((TextEmbeddingModelConfig) modelConfig).getQueryPrefix() == null) {
+                throw new IllegalArgumentException(
+                    "When passing AsymmetricTextEmbeddingParameters, the model requires to be "
+                        + "registered with at least one of `query_prefix` or `passage_prefix`."
+                );
+            }
+            // Passed all checks
+            return true;
+        } else if (mlParams != null) {
+            // mlParams is not null and not an instance of AsymmetricTextEmbeddingParameters.
+            // Should never happen.
+            throw new IllegalArgumentException("TEXT_EMBEDDING algorithm only supports AsymmetricTextEmbeddingParameters.");
+        }
+
+        // mlParams is null, but the model is asymmetric.
+        if (modelConfig != null
+            && (((TextEmbeddingModelConfig) modelConfig).getPassagePrefix() != null
+                || ((TextEmbeddingModelConfig) modelConfig).getQueryPrefix() != null)) {
+            throw new IllegalArgumentException(
+                "The embedding model chosen is asymmetric. To use it, you must declare whether the input is of type `QUERY` or of type `PASSAGE`."
+            );
+        }
+
+        return false;
+    }
+
+    private TextDocsInputDataSet addPrefixesToData(AsymmetricTextEmbeddingParameters mlParams, TextDocsInputDataSet inputDataSet) {
+        // Asymmetric embedding models typically work with "mini-prompts" that prime the model to embed a text
+        // as a query or as a passage. Here we apply the prompt as defined in the model configuration. We default
+        // to passage embedding.
```

Comment:
may be we should default to query embedding.

Reply:
I don't mind to change this. But why do you think this is better?

Reply:
I think query will happen more often than passage?

Reply:
Okay...